### ISSUE TITLE
Inconsistent COPY --dir behaviour

### ISSUE THREAD

**Issue by hbenl** [earthly/earthly#705]
In some situations, `COPY --dir` seems to ignore the `--dir` option or applies it only to some of the source arguments.
To reproduce, create the following file structure:
```
.
├── Earthfile
└── in
    ├── bar
    │   └── b.txt
    └── foo
        └── f.txt
```
with the following Earthfile:
```
FROM debian:buster-slim
WORKDIR /build

test:
  COPY --dir in copied
  RUN apt update && apt install tree && tree
```

Here's the result with different `COPY` lines (and sometimes an additional `RUN` line):
* this ignores the `--dir` argument entirely:
```
COPY --dir in copied
               +test | .
               +test | `-- copied
               +test |     |-- bar
               +test |     |   `-- b.txt
               +test |     `-- foo
               +test |         `-- f.txt
```
* but this works:
```
COPY --dir in copied/
              +test | .
               +test | `-- copied
               +test |     `-- in
               +test |         |-- bar
               +test |         |   `-- b.txt
               +test |         `-- foo
               +test |             `-- f.txt
```
* it also works if the target directory already exists:
```
RUN mkdir copied
COPY --dir in copied 
               +test | .
               +test | `-- copied
               +test |     `-- in
               +test |         |-- bar
               +test |         |   `-- b.txt
               +test |         `-- foo
               +test |             `-- f.txt
```
* this applies `--dir` to the second argument only:
```
COPY --dir in/foo in/bar copied
               +test | .
               +test | `-- copied
               +test |     |-- bar
               +test |     |   `-- b.txt
               +test |     `-- f.txt
```
* again, this can be fixed by adding a trailing `/`:
```
COPY --dir in/foo in/bar copied/
               +test | .
               +test | `-- copied
               +test |     |-- bar
               +test |     |   `-- b.txt
               +test |     `-- foo
               +test |         `-- f.txt
```
* or by creating the directory before copying:
```
RUN mkdir copied
COPY --dir in/foo in/bar copied 
               +test | .
               +test | `-- copied
               +test |     |-- bar
               +test |     |   `-- b.txt
               +test |     `-- foo
               +test |         `-- f.txt
```

**Comment by vladaionescu** [comment]
The intention for `COPY --dir` is that it behaves similarly to a `cp -r`, with the key distinction that `COPY --dir` also creates the parent destination dirs if necessary.

The trailing `/` is meant to signal Earthly that the destination is a directory where the contents should be placed in.

So all the examples you provided work as intended, with the exception of one. And that is

```
COPY --dir in/foo in/bar copied
               +test | .
               +test | `-- copied
               +test |     |-- bar
               +test |     |   `-- b.txt
               +test |     `-- f.txt
```

This is indeed inconsistent as the two source parameters are treated differently.

**Comment by vladaionescu** [comment]
This is now released in `v0.4.5` and reference docs now include some examples about the behavior. Any feedback is appreciated!
